source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+                                         const void *) > callback_function
+            = std_cxx11::bind(&aspect::Particle::World<dim>::load_tracers,
+                              std_cxx11::ref(*this),
+                              std_cxx11::_1,
+                              std_cxx11::_2,
+                              std_cxx11::_3);
+
+          triangulation.notify_ready_to_unpack(data_offset,callback_function);
+          data_offset = numbers::invalid_unsigned_int;
+        }
+    }
+
+    template <int dim>
+    unsigned int
+    World<dim>::cell_weight(const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+                            const typename parallel::distributed::Triangulation<dim>::CellStatus status)
+    {
+      if (cell->active() && !cell->is_locally_owned())
+        return 0;
+
+      if (status == parallel::distributed::Triangulation<dim>::CELL_PERSIST
+          || status == parallel::distributed::Triangulation<dim>::CELL_REFINE)
+        {
+          const types::LevelInd found_cell = std::make_pair<int, int> (cell->level(),cell->index());
+          const unsigned int n_particles_in_cell = particles.count(found_cell);
+          return n_particles_in_cell * tracer_weight;
+        }
+      else if (status == parallel::distributed::Triangulation<dim>::CELL_COARSEN)
+        {
+          unsigned int n_particles_in_cell = 0;
+
+          for (unsigned int child_index = 0; child_index < cell->number_of_children(); ++child_index)
+            {
+              const typename parallel::distributed::Triangulation<dim>::cell_iterator child = cell->child(child_index);
+              const types::LevelInd found_cell = std::make_pair<int, int> (child->level(),child->index());
+              n_particles_in_cell += particles.count(found_cell);
+            }
+          return n_particles_in_cell * tracer_weight;
+        }
+
+      return 0;
+    }
+
+    template <int dim>
+    void
+    World<dim>::store_tracers(const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+                              const typename parallel::distributed::Triangulation<dim>::CellStatus status,
+                              void *data)
+    {
+      unsigned int n_particles_in_cell(0);
+
+      if (status == parallel::distributed::Triangulation<dim>::CELL_PERSIST
+          || status == parallel::distributed::Triangulation<dim>::CELL_REFINE)
+        {
+          const types::LevelInd found_cell = std::make_pair<int, int> (cell->level(),cell->index());
+          const std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::iterator, typename std::multimap<types::LevelInd, Particle<dim> >::iterator> particles_in_cell
+            = particles.equal_range(found_cell);
+          n_particles_in_cell = std::distance(particles_in_cell.first,particles_in_cell.second);
+
+          unsigned int *ndata = static_cast<unsigned int *> (data);
+          *ndata++ = n_particles_in_cell;
```

Comment:
move the increment of the pointer to the next line to make it easier to read?

In cases such as this, I always worry whether `data` will actually have the right alignment for us to store integers or doubles at this location. We're not likely going to encounter any problems because all x86 architectures allow reads and writes from unaligned locations, but that's not true for all processors. We may at one point want to change this sort of thing here to either use `memcpy`, or to just serialize all of the data into a binary blob and then just copy the binary blob to `*data`.